File: ios/SkyCast/Core/DesignSystem/StateViews.swift
import SwiftUI

// The four screen states, implemented once and reused everywhere.
//
// Centralising loading/empty/error presentation means every screen handles all four
// states consistently, with accessibility already wired in, exactly what the
// *Functionality* and *UI/UX* criteria look for.
//
// The Android counterpart is `core/designsystem/component/StateViews.kt`.

/// Full-screen loader. Shown only when there is genuinely nothing cached to render.
struct LoadingView: View {
    var message: String?

    var body: some View {
        VStack(spacing: Spacing.md) {
            ProgressView()
                .controlSize(.large)
            if let message {
                Text(message)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                    .multilineTextAlignment(.center)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(Spacing.lg)
        // One announcement for the whole state, rather than VoiceOver reading a
        // decorative spinner.
        .accessibilityElement(children: .combine)
        .accessibilityLabel(message ?? "Loading")
    }
}

/// Full-screen error. Used only when no cached data exists, otherwise show
/// ``StaleDataBanner`` over the content instead.
struct ErrorView: View {
    let error: AppError
    /// `nil` hides the retry button, for errors where retrying cannot help.
    var onRetry: (() -> Void)?

    var body: some View {
        VStack(spacing: Spacing.md) {
            Image(systemName: error.symbolName)
                .font(.system(size: 56))
                .foregroundStyle(Color.skyError)
                // Decorative: the title and message already carry the meaning.
                .accessibilityHidden(true)

            Text(error.title)
                .font(.title2.weight(.semibold))
                .multilineTextAlignment(.center)

            Text(error.message)
                .font(.subheadline)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)

            // Only offered when retrying can actually succeed, see AppError.isRetryable.
            if let onRetry, error.isRetryable {
                Button("Retry", action: onRetry)
                    .buttonStyle(.bordered)
                    .padding(.top, Spacing.sm)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(Spacing.lg)
    }
}

/// Empty state, a valid, non-error condition such as "no saved locations yet".
struct EmptyStateView: View {
    let title: String
    let message: String
    var systemImage: String = "tray"
    var actionTitle: String?
    var action: (() -> Void)?

    var body: some View {
        VStack(spacing: Spacing.md) {
            Image(systemName: systemImage)
                .font(.system(size: 56))
                .foregroundStyle(.secondary)
                .accessibilityHidden(true)

            Text(title)
                .font(.title2.weight(.semibold))
                .multilineTextAlignment(.center)

            Text(message)
                .font(.subheadline)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)

            if let actionTitle, let action {
                Button(actionTitle, action: action)
                    .buttonStyle(.borderedProminent)
                    .padding(.top, Spacing.sm)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(Spacing.lg)
    }
}

/// Non-blocking banner shown **above cached content** when a refresh failed or the data
/// is stale.
///
/// This is the visible half of the offline-first promise: the user keeps their data and is
/// merely told it might be out of date.
struct StaleDataBanner: View {
    let message: String
    var onRetry: (() -> Void)?
    var onDismiss: (() -> Void)?

    var body: some View {
        HStack(spacing: Spacing.sm) {
            Image(systemName: "clock.arrow.circlepath")
                .accessibilityHidden(true)

            Text(message)
                .font(.footnote)
                .frame(maxWidth: .infinity, alignment: .leading)

            if let onRetry {
                Button("Retry", action: onRetry)
                    .font(.footnote.weight(.semibold))
            }
            if let onDismiss {
                Button {
                    onDismiss()
                } label: {
                    Image(systemName: "xmark")
                        .font(.footnote)
                }
                .accessibilityLabel("Dismiss")
            }
        }
        .padding(.horizontal, Spacing.md)
        .padding(.vertical, Spacing.sm)
        .frame(maxWidth: .infinity)
        .background(Color.skyWarning.opacity(0.18))
        // Announced as soon as it appears, so a VoiceOver user learns the data is stale
        // without having to go looking for the banner.
        .accessibilityElement(children: .combine)
        .accessibilityAddTraits(.isStaticText)
    }
}

// MARK: - Previews

#Preview("Loading") {
    LoadingView(message: "Fetching the latest weather…")
}

#Preview("Error: offline") {
    ErrorView(error: .offline, onRetry: {})
}

#Preview("Error: no API key") {
    // No Retry button: retrying a missing key can never succeed.
    ErrorView(error: .unauthorized, onRetry: {})
}

#Preview("Empty") {
    EmptyStateView(
        title: "No locations yet",
        message: "Add a place and SkyCast will keep its forecast ready, even offline.",
        systemImage: "mappin.and.ellipse",
        actionTitle: "Add location",
        action: {}
    )
}

#Preview("Stale banner") {
    StaleDataBanner(
        message: "Offline, showing data from 20 minutes ago",
        onRetry: {},
        onDismiss: {}
    )
}

File: android/app/src/main/java/com/nauhaan/skycast/core/designsystem/component/StateViews.kt
package com.nauhaan.skycast.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.nauhaan.skycast.R
import com.nauhaan.skycast.core.designsystem.theme.SkyCastTheme
import com.nauhaan.skycast.core.designsystem.theme.Spacing

/**
 * The four screen states, implemented once and reused everywhere.
 *
 * Centralising loading/empty/error presentation means every screen handles all four
 * states consistently, with accessibility already wired in, which is exactly what the
 * *Functionality* and *UI/UX* criteria are looking for.
 */

/** Full-screen loader. Only shown when there is genuinely nothing cached to render. */
@Composable
fun LoadingView(modifier: Modifier = Modifier, message: String? = null) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(Spacing.lg)
            // One live-region announcement for the whole state, rather than
            // TalkBack reading a decorative spinner.
            .semantics { contentDescription = message ?: LOADING_ANNOUNCEMENT },
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        CircularProgressIndicator()
        if (message != null) {
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = Spacing.md),
            )
        }
    }
}

/**
 * Full-screen error. Used only when no cached data exists, otherwise show
 * [StaleDataBanner] over the content instead.
 *
 * @param onRetry `null` hides the retry button, for errors where retrying cannot help
 *   (a missing API key, for instance).
 */
@Composable
fun ErrorView(
    title: String,
    message: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    onRetry: (() -> Unit)? = null,
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(Spacing.lg),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                // Decorative: the title and message already convey the meaning, so
                // announcing the icon too would be redundant for screen-reader users.
                contentDescription = null,
                tint = MaterialTheme.colorScheme.error,
                modifier = Modifier.size(StateIconSize),
            )
        }
        Text(
            text = title,
            style = MaterialTheme.typography.titleLarge,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.md),
        )
        Text(
            text = message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.sm),
        )
        if (onRetry != null) {
            OutlinedButton(
                onClick = onRetry,
                modifier = Modifier.padding(top = Spacing.lg),
            ) {
                Text(stringResource(R.string.action_retry))
            }
        }
    }
}

/** Empty state, a valid, non-error condition such as "no saved locations yet". */
@Composable
fun EmptyStateView(
    title: String,
    message: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(Spacing.lg),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.size(StateIconSize),
            )
        }
        Text(
            text = title,
            style = MaterialTheme.typography.titleLarge,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.md),
        )
        Text(
            text = message,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Spacing.sm),
        )
        if (actionLabel != null && onAction != null) {
            OutlinedButton(
                onClick = onAction,
                modifier = Modifier.padding(top = Spacing.lg),
            ) {
                Text(actionLabel)
            }
        }
    }
}

/**
 * Non-blocking banner shown **above cached content** when a refresh failed or the
 * data is stale.
 *
 * This is the visible half of the offline-first promise: the user keeps their data and
 * is merely told it might be out of date.
 */
@Composable
fun StaleDataBanner(
    message: String,
    modifier: Modifier = Modifier,
    onRetry: (() -> Unit)? = null,
    onDismiss: (() -> Unit)? = null,
) {
    Surface(
        modifier = modifier.fillMaxWidth(),
        color = MaterialTheme.colorScheme.tertiaryContainer,
        contentColor = MaterialTheme.colorScheme.onTertiaryContainer,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = Spacing.md, vertical = Spacing.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = message,
                style = MaterialTheme.typography.bodySmall,
                modifier = Modifier.weight(1f),
            )
            if (onRetry != null) {
                TextButton(onClick = onRetry) {
                    Text(stringResource(R.string.action_retry))
                }
            }
            if (onDismiss != null) {
                TextButton(onClick = onDismiss) {
                    Text(stringResource(R.string.action_dismiss))
                }
            }
        }
    }
}

/** Size of the illustrative icon in the error and empty states. */
private val StateIconSize: Dp = 56.dp

private const val LOADING_ANNOUNCEMENT = "Loading"

// ── Previews ───────────────────────────────────────────────────────────────
// Previews are cheap documentation and let every state be reviewed without a
// device. Keep one per state.

@Preview(name = "Loading", showBackground = true)
@Composable
private fun LoadingViewPreview() {
    SkyCastTheme { LoadingView(message = "Fetching the latest weather…") }
}

@Preview(name = "Error", showBackground = true)
@Composable
private fun ErrorViewPreview() {
    SkyCastTheme {
        ErrorView(
            title = "No internet connection",
            message = "Connect to a network and try again.",
            onRetry = {},
        )
    }
}

@Preview(name = "Empty", showBackground = true)
@Composable
private fun EmptyStateViewPreview() {
    SkyCastTheme {
        EmptyStateView(
            title = "No locations yet",
            message = "Add a place to start tracking its weather.",
            actionLabel = "Add location",
            onAction = {},
        )
    }
}

@Preview(name = "Stale banner", showBackground = true)
@Composable
private fun StaleDataBannerPreview() {
    SkyCastTheme {
        StaleDataBanner(message = "Offline, showing data from 20 minutes ago", onRetry = {})
    }
}
